// @brendangooch/simple-game-loop/lib/simple-game-loop.js
export class SimpleGameLoop {
    constructor() {
        this.objectsToUpdate = [];
        this.running = false;
        this.RAF = null;
        this.previous = 0;
        this.ms = 0;
    }
    add(obj) {
        this.objectsToUpdate.push(obj);
    }
    start() {
        if (!this.running) {
            this.bindRAF();
            this.previous = 0;
            this.running = true;
        }
    }
    stop() {
        if (this.running) {
            if (this.RAF)
                cancelAnimationFrame(this.RAF);
            this.RAF = null;
            this.running = false;
        }
    }
    update() {
        this.updateMS();
        this.objectsToUpdate.forEach(obj => obj.update(this.ms));
        this.bindRAF();
    }
    updateMS() {
        const now = Date.now();
        if (!this.previous) {
            this.previous = now;
        }
        this.ms = now - this.previous;
        this.previous = now;
    }
    bindRAF() {
        this.RAF = requestAnimationFrame(this.update.bind(this));
    }
}
