// @brendangooch/canvas/lib/canvas-shadow.js
import AbstractCanvasComponent from "./abstract-canvas-component.js";
export default class CanvasShadow extends AbstractCanvasComponent {
    constructor() {
        super(...arguments);
        this.props = {
            x: 0,
            y: 0,
            blur: 0,
            color: 'rgba(0,0,0,0)'
        };
    }
    x(n) {
        this.props.x = n;
        this.update();
        return this;
    }
    y(n) {
        this.props.y = n;
        this.update();
        return this;
    }
    blur(n) {
        this.props.blur = n;
        this.update();
        return this;
    }
    color(c) {
        this.props.color = c;
        this.update();
        return this;
    }
    reset() {
        this.props.x = 0;
        this.props.y = 0;
        this.props.blur = 0;
        this.props.color = 'rgba(0,0,0,0)';
        this.update();
    }
    update() {
        this.ctx.shadowOffsetX = this.props.x;
        this.ctx.shadowOffsetY = this.props.y;
        this.ctx.shadowBlur = this.props.blur;
        this.ctx.shadowColor = this.props.color;
    }
}
