// @brendangooch/dynamic-objects/lib/base-dynamic-object.js
export class BaseDynamicObject {
    constructor() {
        this._duration = 0;
        this.easeOption = 'noEase';
        this.isOn = false;
    }
    duration(ms) {
        if (!this.isActive && ms > 0)
            this._duration = ms;
    }
    ease(easeOption) {
        if (!this.isActive)
            this.easeOption = easeOption;
    }
    update(ms) {
        if (this.isOn && this.isActive) {
            this.increment(ms);
            this.updateCurrent();
            if (!this.isActive)
                this.updateComplete();
        }
    }
    save() {
        return JSON.stringify({
            isOn: this.isOn,
            duration: this._duration,
            easeOption: this.easeOption
        });
    }
    load(json) {
        const state = JSON.parse(json);
        if (state.isOn === undefined)
            return false;
        if (state.duration === undefined)
            return false;
        if (state.easeOption === undefined)
            return false;
        this.isOn = state.isOn;
        this._duration = state.duration;
        this.easeOption = state.easeOption;
        return true;
    }
    turnOn() {
        this.isOn = true;
    }
    turnOff() {
        this.isOn = false;
    }
}
