// @brendangooch/dynamic-objects/lib/dynamic-string.js
import { BaseDynamicObject } from "./base-dynamic-object.js";
import { DynamicNumber } from "./dynamic-number.js";
export class DynamicString extends BaseDynamicObject {
    constructor() {
        super(...arguments);
        this.string = '';
        this.currentValue = '';
        this.index = new DynamicNumber();
    }
    get isActive() {
        return this.index.isActive;
    }
    get current() {
        return this.currentValue;
    }
    duration(ms) {
        super.duration(ms);
        return this;
    }
    ease(easeOption) {
        super.ease(easeOption);
        return this;
    }
    changeTo(next) {
        if (this.canChange(next))
            return this.doChange(next);
        return 0;
    }
    save() {
        return JSON.stringify({
            parent: super.save(),
            string: this.string,
            currentValue: this.currentValue,
            index: this.index.save()
        });
    }
    load(json) {
        const state = JSON.parse(json);
        if (state.parent === undefined)
            return false;
        if (state.string === undefined)
            return false;
        if (state.currentValue === undefined)
            return false;
        if (state.index === undefined)
            return false;
        const parentLoaded = super.load(state.parent);
        this.string = state.string;
        this.currentValue = state.currentValue;
        this.index.load(state.index);
        return parentLoaded;
    }
    increment(ms) {
        this.index.update(ms);
    }
    updateCurrent() {
        this.currentValue = this.string.substring(0, this.index.rounded);
    }
    updateComplete() {
        this.reset();
        this.turnOff();
    }
    canChange(next) {
        return !this.isActive && next !== '';
    }
    doChange(next) {
        this.string = next;
        if (this._duration > 0)
            this.dynamicChange();
        else
            this.instantChange();
        return this._duration;
    }
    instantChange() {
        this.currentValue = this.string;
    }
    dynamicChange() {
        this.currentValue = '';
        this.index.duration(this._duration).ease(this.easeOption).changeTo(this.string.length);
        this.turnOn();
    }
    reset() {
        this.index.changeTo(0);
        this._duration = 0;
        this.easeOption = 'noEase';
    }
}
DynamicString.MAX_CHARS = 1000;
