// @brendangooch/maths/lib/cubic-bezier-curve.js
import { Vector2D } from "./vector-2d.js";
export class CubicBezierCurve {
    constructor() {
        this.s = new Vector2D(0, 0);
        this.a = new Vector2D(0, 0);
        this.b = new Vector2D(0, 0);
        this.e = new Vector2D(0, 0);
        this.equal = true;
    }
    x(t) {
        if (t === 0)
            return this.s.x;
        if (t === 1)
            return this.e.x;
        if (this.equal)
            return this.s.x;
        return this.calculateX(t);
    }
    y(t) {
        if (t === 0)
            return this.s.y;
        if (t === 1)
            return this.e.y;
        if (this.equal)
            return this.s.y;
        return this.calculateY(t);
    }
    setStart(x, y) {
        this.s.setXY(x, y);
        this.checkEqual();
        return this;
    }
    setControlA(x, y) {
        this.a.setXY(x, y);
        this.checkEqual();
        return this;
    }
    setControlB(x, y) {
        this.b.setXY(x, y);
        this.checkEqual();
        return this;
    }
    setEnd(x, y) {
        this.e.setXY(x, y);
        this.checkEqual();
        return this;
    }
    setAll(x, y) {
        this.setStart(x, y);
        this.setControlA(x, y);
        this.setControlB(x, y);
        this.setEnd(x, y);
    }
    distanceBetween() {
        return this.s.distanceTo(this.e);
    }
    save() {
        return JSON.stringify({
            start: this.s.save(),
            controlA: this.a.save(),
            controlB: this.b.save(),
            end: this.e.save(),
            equal: this.equal,
        });
    }
    load(json) {
        const state = JSON.parse(json);
        if (state.start === undefined)
            throw new Error('missing "start" property');
        if (state.controlA === undefined)
            throw new Error('missing "controlA" property');
        if (state.controlB === undefined)
            throw new Error('missing "controlB" property');
        if (state.end === undefined)
            throw new Error('missing "end" property');
        if (state.equal === undefined)
            throw new Error('missing "equal" property');
        this.s.load(state.start);
        this.a.load(state.controlA);
        this.b.load(state.controlB);
        this.e.load(state.end);
        this.equal = state.equal;
    }
    calculateX(t) {
        return Math.pow(1 - t, 3) * this.s.x + Math.pow(1 - t, 2) * 3 * t * this.a.x + (1 - t) * 3 * t * t * this.b.x + t * t * t * this.e.x;
    }
    calculateY(t) {
        return Math.pow(1 - t, 3) * this.s.y + Math.pow(1 - t, 2) * 3 * t * this.a.y + (1 - t) * 3 * t * t * this.b.y + t * t * t * this.e.y;
    }
    checkEqual() {
        this.equal = this.s.equals(this.a) && this.s.equals(this.b) && this.s.equals(this.e);
    }
}
