// @brendangooch/playing-card/lib/rank.js
import {} from "./index.js";
import { getRankFromType } from "./get-rank-from-type.js";
export class Rank {
    constructor(type) {
        this.type = type;
    }
    get number() {
        return getRankFromType(this.type);
    }
    get name() {
        switch (this.number) {
            case 0:
                return 'joker';
            case 1:
                return 'ace';
            case 2:
                return 'two';
            case 3:
                return 'three';
            case 4:
                return 'four';
            case 5:
                return 'five';
            case 6:
                return 'six';
            case 7:
                return 'seven';
            case 8:
                return 'eight';
            case 9:
                return 'nine';
            case 10:
                return 'ten';
            case 11:
                return 'jack';
            case 12:
                return 'queen';
            case 13:
                return 'king';
        }
        return 'undefined';
    }
    get nameDefinite() {
        return `${this.definiteArticle} ${this.name}`;
    }
    get nameIndefinite() {
        return `the ${this.name}`;
    }
    get namePlural() {
        return `${this.name}${this.pluralEnding}`;
    }
    get definiteArticle() {
        return (this.number === 1 || this.number === 8) ? 'an' : 'a';
    }
    get pluralEnding() {
        return (this.number === 6) ? 'es' : 's';
    }
}
