// @brendangooch/maths/lib/quadratic-bezier-curve.js
import { clamp } from "./clamp.js";
export class QuadraticBezierCurve {
    constructor(start, control, end) {
        this.start = start.clone();
        this.control = control.clone();
        this.end = end.clone();
    }
    save() {
        return JSON.stringify({
            start: this.start.save(),
            control: this.control.save(),
            end: this.end.save()
        });
    }
    load(json) {
        const state = JSON.parse(json);
        if (state.start === undefined)
            throw new Error('missing "start" property');
        if (state.control === undefined)
            throw new Error('missing "control" property');
        if (state.end === undefined)
            throw new Error('missing "end" property');
        this.start.load(state.start);
        this.control.load(state.control);
        this.end.load(state.end);
    }
    x(t) {
        t = clamp(t, 0, 1);
        return Math.pow(1 - t, 2) * this.start.x + (1 - t) * 2 * t * this.control.x + t * t * this.end.x;
    }
    y(t) {
        t = clamp(t, 0, 1);
        return Math.pow(1 - t, 2) * this.start.y + (1 - t) * 2 * t * this.control.y + t * t * this.end.y;
    }
}
