// @brendangooch/maths/lib/vector-2d.js
import { distanceBetween } from "./distance-between.js";
export class Vector2D {
    constructor(x = 1, y = 0) {
        this._x = x;
        this._y = y;
    }
    get x() {
        return this._x;
    }
    set x(x) {
        this._x = x;
    }
    get y() {
        return this._y;
    }
    set y(y) {
        this._y = y;
    }
    get angle() {
        return Math.atan2(this._y, this._x);
    }
    set angle(radians) {
        const length = this.length;
        this._x = Math.cos(radians) * length;
        this._y = Math.sin(radians) * length;
    }
    get length() {
        return Math.sqrt(this._x * this._x + this._y * this._y);
    }
    set length(px) {
        const angle = this.angle;
        this._x = Math.cos(angle) * px;
        this._y = Math.sin(angle) * px;
    }
    add(v2) {
        return new Vector2D(this._x + v2.x, this._y + v2.y);
    }
    subtract(v2) {
        return new Vector2D(this._x - v2.x, this._y - v2.y);
    }
    multiply(val) {
        return new Vector2D(this._x * val, this._y * val);
    }
    divide(val) {
        return new Vector2D(this._x / val, this._y / val);
    }
    addTo(v2) {
        this._x += v2.x;
        this._y += v2.y;
    }
    subtractFrom(v2) {
        this._x -= v2.x;
        this._y -= v2.y;
    }
    multiplyBy(val) {
        this._x *= val;
        this._y *= val;
    }
    divideBy(val) {
        this._x /= val;
        this._y /= val;
    }
    distanceTo(v2) {
        return distanceBetween(this._x, this._y, v2.x, v2.y);
    }
    normalize() {
        this.divideBy(this.length);
    }
    copy(v2) {
        this._x = v2.x;
        this._y = v2.y;
    }
    clone() {
        return new Vector2D(this._x, this._y);
    }
    save() {
        return JSON.stringify({
            x: this._x,
            y: this._y
        });
    }
    load(json) {
        const state = JSON.parse(json);
        if (state.x === undefined)
            throw new Error('missing "x" property');
        if (state.y === undefined)
            throw new Error('missing "y" property');
        this.x = state.x;
        this.y = state.y;
    }
}
